### ISSUE TITLE
0137. Single Number I I | LeetCode Cookbook

### ISSUE THREAD

**Issue by halfrost** [halfrost/S2#458]
https://books.halfrost.com/leetcode/ChapterFour/0100~0199/0137.Single-Number-II/

**Comment by mrzz00** [comment]
在singleNumberII中
> twos = (twos ^ nums[i]) & ^ones
用真值表检验这句表达式的正确性，但我理解不了`twos & num s[i] `的含义诶？？

**Comment by halfrost** [comment]
> 在singleNumberII中
> 
> > twos = (twos ^ nums[i]) & ^ones
> > 用真值表检验这句表达式的正确性，但我理解不了`twos & num s[i] `的含义诶？？

代码中没有 twos & num s[i]，你想问的是 twos ^ nums[i] 吧？异或操作为的是统计每个数位上的 1 。x^x=0，x^0=x

**Comment by mrzz00** [comment]
> @halfrost
> > 在singleNumberII中
> > 
> > > twos = (twos ^ nums[i]) & ^ones
> > > 用真值表检验这句表达式的正确性，但我理解不了`twos & num s[i] `的含义诶？？
> 
> @mrzz00 代码中没有 twos & num s[i]，你想问的是 twos ^ nums[i] 吧？异或操作为的是统计每个数位上的 1 。x^x=0，x^0=x


> @halfrost
> > 在singleNumberII中
> > 
> > > twos = (twos ^ nums[i]) & ^ones
> > > 用真值表检验这句表达式的正确性，但我理解不了`twos & num s[i] `的含义诶？？
> 
> @mrzz00 代码中没有 twos & num s[i]，你想问的是 twos ^ nums[i] 吧？异或操作为的是统计每个数位上的 1 。x^x=0，x^0=x

抱歉出现了手误。我明白可以用异或来统计不同位的1，就像  ones^nums[i]。
而twos里面出现的记录1是因为nums[i]和ones相同位都是1，twos|(ones&nums[i])；
消除twos里面出现的记录1是因为ones和twos相同位都是1，twos& ^(ones&twos)
所以不理解为什么仅仅twos^nums[i]就可以统计目前出现过两次的比特位？

**Comment by halfrost** [comment]
@mrzz00 不应该说 twos 里面统计的是出现 2 次的比特位。x^x = 0，用一个变量即可消去出现 2 次的数字。此题需要消去出现 3 次的数，即 3 进制进位，00，01，10 。用 (twos, ones) 这样 2 位统计数位上 1 出现的次数。你重新刷新一下本页面，我把表格画出来了。你对照 0，1 表格看看。ones 初始为 0，+1 以后变成 1，再 +1 需要进 1 位到 twos 位上。twos 由 0 变成 1，再 + 1，ones 和 twos 均变 0 。每疑惑一个 A[i]，(twos,ones) 需要对应变化。

**Comment by mrzz00** [comment]
> @halfrost
> @mrzz00 不应该说 twos 里面统计的是出现 2 次的比特位。x^x = 0，用一个变量即可消去出现 2 次的数字。此题需要消去出现 3 次的数，即 3 进制进位，00，01，10 。用 (twos, ones) 这样 2 位统计数位上 1 出现的次数。你重新刷新一下本页面，我把表格画出来了。你对照 0，1 表格看看。ones 初始为 0，+1 以后变成 1，再 +1 需要进 1 位到 twos 位上。twos 由 0 变成 1，再 + 1，ones 和 twos 均变 0 。每疑惑一个 A[i]，(twos,ones) 需要对应变化。

我明白了，按进制来理解的话，表达式的含义就很清楚了。所以one和two其实代表的权重，就像5次数字里面的three其实指的是出现过4次的比特位。谢谢你的耐心讲解！

**Comment by halfrost** [comment]
@mrzz00 是的。是按照进制来理解的。原来最初是出现 2 次就消除，x^x = 0，现在出现 3 次的数要消除，那么需要用 2 位数字来表示消除中的状态，00，01，10。(twos,ones) 这个组合代表的意义就是这样，ones 是进制的低位，twos 是进制的高位。如果是出现 5 次消除，那么需要 3 位二进制数来表示，于是用到了 (threes,twos,ones) 这样的组合。每一位的更新逻辑也可以仿造 3 次消除，先更新 threes，再更新 twos，最后更新 ones。至于如何更新，画一个真值表观察一下就行。这道题我把真值表的部分也更新在题解里面了，你刷新一下页面可以看到。

**Comment by mrzz00** [comment]
> @halfrost
> @mrzz00 是的。是按照进制来理解的。原来最初是出现 2 次就消除，x^x = 0，现在出现 3 次的数要消除，那么需要用 2 位数字来表示消除中的状态，00，01，10。(twos,ones) 这个组合代表的意义就是这样，ones 是进制的低位，twos 是进制的高位。如果是出现 5 次消除，那么需要 3 位二进制数来表示，于是用到了 (threes,twos,ones) 这样的组合 000，001，010，011，100。每一位的更新逻辑也可以仿造 3 次消除，先更新 threes，再更新 twos，最后更新 ones。至于如何更新，画一个真值表观察一下就行。这道题我把真值表的部分也更新在题解里面了，你刷新一下页面可以看到。

看到啦，读真值表就一目了然，然后推演一下表达式就得到了结果。多谢！
